lib/install/management/appliance.go

```diff
@@ -181,19 +181,39 @@ func (d *Dispatcher) deleteVM(vm *vm.VirtualMachine, force bool) error {
 		}
 	}
 
-	// Only retry VM destroy on CurrentAccess error
+	retryRemovalError := func(err error) bool {
+		return tasks.IsConcurrentAccessError(err) || tasks.IsInvalidPowerStateError(err)
+	}
+
+	// Only retry VM destroy on ConcurrentAccess error
+	var destroyFirstAttempt bool
 	err = retry.Do(func() error {
+		if !destroyFirstAttempt {
+			d.op.Debugf("detected an OOB powerOn for container (%s) attempting an additional shutdown before removal", vm.Reference().String())
+			vm.PowerOff(d.op)
+		}
+		destroyFirstAttempt = true
+
 		_, err := vm.DeleteExceptDisks(d.op)
 		return err
-	}, tasks.IsConcurrentAccessError)
+	}, retryRemovalError)
 
 	if err != nil {
 		d.op.Warnf("Destroy VM %s failed with %s, unregister the VM instead", vm.Reference(), err)
 
-		// Only retry VM unregister on ConcurrentAccess error
+		var unregisterFirstAttempt bool
+		// Only retry VM unregister on InvalidPowerState error, This will allow us to avoid OOB power ups of target vms.
 		err = retry.Do(func() error {
+			if !unregisterFirstAttempt {
+				d.op.Debugf("detected an OOB powerOn for container (%s) attempting an additional shutdown before unregister", vm.Reference().String())
+				// NOTE: ignore the error here. If it is a RuntimeFault then the following unregister will fail as well.
+				vm.PowerOff(d.op)
+			}
+			unregisterFirstAttempt = true
+
 			return vm.Unregister(d.op)
-		}, tasks.IsConcurrentAccessError)
+			// TODO: retry on InvalidPowerState. Unregister does not have the potential to return a ConcurrentAccess return
```

Comment:
If `vm.Unregister` can't return a concurrent access error, what about `vm.PowerOff`?

Reply:
I need to remove that comment. In one of the linked tickets there was a log that actually showed us getting a concurrent access error from `vm.unregister` oddly enough that is not listed in the vsphere documentation. I suspect the type for that error return does not show up as `ConcurrentAccessFault`

Reply:
@zjs so `vm.PowerOff` will return an `InvalidPowerState` fault if the vm is in any other powerstate than `PoweredOn` so it can. But I do not think we should be concerned if the vm.PowerOff returns an InvalidPowerstate so much. We will when we move to doing the additional poweroff in the retry certifier function(comment from @hickeng below). We do need to be careful that we are not retrying on InvalidPowerstate from a PowerOff as it indicates the opposite case of `destroy` and `unregister` and that would then cause a infinite retry loop(until timeout). I was trying to be careful to make sure we did not do that. I will make sure that is still the case after I respond to everyone's comments.